### PR TITLE
feat(grpo): mask env-flagged samples from policy loss

### PR DESCRIPTION
**Pull request by macandro96**
# What does this PR do ?

Adds a per-sample gradient mask driven by the rollout environment. When `instance_config.mask_sample` is `True` on a rollout result, the sample's `loss_multiplier` is zeroed out before the policy update — but the sample is kept for advantage computation (baseline / std).

Cherry-pick commit [here](https://gitlab-master.nvidia.com/terryk/nemo-rl-internal/-/merge_requests/222)

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by macandro96** [comment]
/ok to test [c112b82](https://github.com/NVIDIA-NeMo/RL/pull/3163/commits/c112b8222fc0f77f0994d4999caa90c147f52b33)

**Comment by macandro96** [comment]
/ok to test [8112e34](https://github.com/NVIDIA-NeMo/RL/commit/8112e345d8ee072dd87aa8561648f8748695c990)

**Review by sdevare-nv — DISMISSED** [review]
looks good to me!

File: nemo_rl/experience/rollouts.py [hunk 1]
```diff
@@ -1974,6 +1989,9 @@ def run_async_nemo_gym_rollout(
             "truncated": torch.tensor(
                 [m["hit_max_tokens"] for m in all_sample_metrics], dtype=torch.bool
             ),
+            # Agent/env-driven mask flag — True means this sample should be masked
+            # from the GRPO gradient (kept for advantage computation).
+            "mask_sample": _extract_mask_sample_flags(results),
```
Comment (yfw) [inline nemo_rl/experience/rollouts.py:1994]: [`rollouts.py:1994`](https://github.com/NVIDIA-NeMo/RL/blob/8112e345d8ee072dd87aa8561648f8748695c990/nemo_rl/experience/rollouts.py#L1994)

**An existing test will break.** This unconditionally adds a `mask_sample` key to `final_batch`, but [`test_run_async_nemo_gym_rollout._standardize`](https://github.com/NVIDIA-NeMo/RL/blob/8112e345d8ee072dd87aa8561648f8748695c990/tests/unit/experience/test_rollouts.py#L1006) (`@pytest.mark.nemo_gym`) pops only `message_log`/`truncated`. `expected_result` has no `mask_sample`, so the dict-equality assert at [test_rollouts.py:1029](https://github.com/NVIDIA-NeMo/RL/blob/8112e345d8ee072dd87aa8561648f8748695c990/tests/unit/experience/test_rollouts.py#L1029) will fail when the `nemo_gym` CI job runs. It doesn't run in default L0, which is why it wasn't caught locally.

Suggest mirroring the `truncated` handling: add `mask_sample` to `expected_result["final_batch"]` and pop + bool-check it in `_standardize`.

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -1811,6 +1811,24 @@ def _build_async_grpo_train_data(
...
 
+def _apply_mask_sample_filter(repeated_batch: BatchedDataDict[DatumSpec]) -> int:
+    """Zero loss_multiplier where mask_sample is True and return the count."""
+    if "mask_sample" not in repeated_batch:
+        return 0
+
+    loss_multiplier = repeated_batch["loss_multiplier"].clone()
+    mask_sample = repeated_batch["mask_sample"]
+
+    if isinstance(mask_sample, list):
```
Comment (yfw) [inline nemo_rl/algorithms/grpo.py:1822]: [`grpo.py:1822`](https://github.com/NVIDIA-NeMo/RL/blob/8112e345d8ee072dd87aa8561648f8748695c990/nemo_rl/algorithms/grpo.py#L1822)

The list-valued `mask_sample` branch here isn't exercised — both new tests pass tensors. A small addition to `TestMaskSampleFilter` (verified passing locally) covers it:

```python
    def test_masks_list_valued_mask_sample(self):
        repeated_batch = BatchedDataDict(
            {
                "loss_multiplier": torch.tensor([1.0, 1.0, 1.0]),
                "mask_sample": [True, False, True],
            }
        )
        num_masked = _apply_mask_sample_filter(repeated_batch)
        assert num_masked == 2
        assert torch.equal(
            repeated_batch["loss_multiplier"], torch.tensor([0.0, 1.0, 0.0])
        )
```

**Comment by macandro96** [comment]
/ok to test [27a0f20](https://github.com/NVIDIA-NeMo/RL/commit/27a0f209fbde7f6e3e3fc8a477e49d69f2fe09dd)

**Review by yfw — APPROVED** [review]
